### PR TITLE
Improve speed of Nearby Look-ups in GridSpaces

### PR DESCRIPTION
**Pull request by Tortar**
This came to my mind today, it's a simple change but has a big impact: 
Flocking in the large version is 1.5x faster now! edit: but with a big catch :D, see below

In general it has a good impact when periodic is true and just a slight impact when it is false.

It's time for 5.15 I guess  :D

### CODE REVIEW

**Comment by Tortar** [comment]
Wait a second, I don't see much of the speed-up I saw earlier, need to investigate a bit better

**Review by Datseris — COMMENTED** [review]
Why does one inequality use `<=` and the other `<` ?  
```julia
(i -> r < pos[i] <= space_size[i] - r, 1:D)
```

**Comment by Tortar** [comment]
Now we can actually evaluate the matter:

```julia
using Agents, BenchmarkTools

@agent A GridAgent{2} begin
end

function periodic_model(GridType; numagents = 1000, griddims = (50, 50))
    space = GridType(griddims, periodic = true)
    model = ABM(A, space)
    for _ in 1:numagents
        add_agent_single!(model)
    end
    return model
end

function aperiodic_model(GridType; numagents = 1000, griddims = (50, 50))
    space = GridType(griddims, periodic = false)
    model = ABM(A, space)
    for _ in 1:numagents
        add_agent_single!(model)
    end
    return model
end

pos = (20,20)

# ids in gridspace
@btime collect(nearby_ids($pos, model, 1)) setup=(model=periodic_model(GridSpace))
@btime collect(nearby_ids($pos, model, 1)) setup=(model=aperiodic_model(GridSpace))
@btime collect(nearby_ids($pos, model, 10)) setup=(model=periodic_model(GridSpace))
@btime collect(nearby_ids($pos, model, 10)) setup=(model=aperiodic_model(GridSpace))

# ids in gridspacesingle
@btime collect(nearby_ids($pos, model, 1)) setup=(model=periodic_model(GridSpaceSingle))
@btime collect(nearby_ids($pos, model, 1)) setup=(model=aperiodic_model(GridSpaceSingle))
@btime collect(nearby_ids($pos, model, 10)) setup=(model=periodic_model(GridSpaceSingle))
@btime collect(nearby_ids($pos, model, 10)) setup=(model=aperiodic_model(GridSpaceSingle))

# positions
@btime collect(nearby_positions($pos, model, 1)) setup=(model=periodic_model(GridSpace))
@btime collect(nearby_positions($pos, model, 1)) setup=(model=aperiodic_model(GridSpace))
@btime collect(nearby_positions($pos, model, 10)) setup=(model=periodic_model(GridSpace))
@btime collect(nearby_positions($pos, model, 10)) setup=(model=aperiodic_model(GridSpace))
```

gives before and after this PR:

```julia
# before PR

julia> # ids in gridspace
       @btime collect(nearby_ids($pos, model, 1)) setup=(model=periodic_model(GridSpace))
  59.522 ns (1 allocation: 64 bytes)

julia> @btime collect(nearby_ids($pos, model, 1)) setup=(model=aperiodic_model(GridSpace))
  42.738 ns (1 allocation: 64 bytes)

julia> @btime collect(nearby_ids($pos, model, 10)) setup=(model=periodic_model(GridSpace))
  4.242 μs (150 allocations: 9.66 KiB)

julia> @btime collect(nearby_ids($pos, model, 10)) setup=(model=aperiodic_model(GridSpace))
  1.843 μs (5 allocations: 2.67 KiB)

julia> # ids in gridspacesingle
       @btime collect(nearby_ids($pos, model, 1)) setup=(model=periodic_model(GridSpaceSingle))
  71.976 ns (1 allocation: 64 bytes)

julia> @btime collect(nearby_ids($pos, model, 1)) setup=(model=aperiodic_model(GridSpaceSingle))
  40.997 ns (1 allocation: 64 bytes)

julia> @btime collect(nearby_ids($pos, model, 10)) setup=(model=periodic_model(GridSpaceSingle))
  3.830 μs (6 allocations: 2.92 KiB)

julia> @btime collect(nearby_ids($pos, model, 10)) setup=(model=aperiodic_model(GridSpaceSingle))
  1.815 μs (6 allocations: 2.74 KiB)

julia> # positions
       @btime collect(nearby_positions($pos, model, 1)) setup=(model=periodic_model(GridSpace))
  83.006 ns (1 allocation: 193 bytes)

julia> @btime collect(nearby_positions($pos, model, 1)) setup=(model=aperiodic_model(GridSpace))
  107.129 ns (3 allocations: 273 bytes)

julia> @btime collect(nearby_positions($pos, model, 10)) setup=(model=periodic_model(GridSpace))
  3.320 μs (2 allocations: 8.61 KiB)

julia> @btime collect(nearby_positions($pos, model, 10)) setup=(model=aperiodic_model(GridSpace))
  3.310 μs (7 allocations: 16.67 KiB)

# after PR

julia> # ids in gridspace
       @btime collect(nearby_ids($pos, model, 1)) setup=(model=periodic_model(GridSpace))
  61.147 ns (1 allocation: 64 bytes)

julia> @btime collect(nearby_ids($pos, model, 1)) setup=(model=aperiodic_model(GridSpace))
  41.979 ns (1 allocation: 64 bytes)

julia> @btime collect(nearby_ids($pos, model, 10)) setup=(model=periodic_model(GridSpace))
  4.212 μs (147 allocations: 9.52 KiB)

julia> @btime collect(nearby_ids($pos, model, 10)) setup=(model=aperiodic_model(GridSpace))
  2.881 μs (144 allocations: 9.27 KiB)

julia> # ids in gridspacesingle
       @btime collect(nearby_ids($pos, model, 1)) setup=(model=periodic_model(GridSpaceSingle))
  42.718 ns (2 allocations: 112 bytes)

julia> @btime collect(nearby_ids($pos, model, 1)) setup=(model=aperiodic_model(GridSpaceSingle))
  42.495 ns (2 allocations: 112 bytes)

julia> @btime collect(nearby_ids($pos, model, 10)) setup=(model=periodic_model(GridSpaceSingle))
  1.646 μs (7 allocations: 2.78 KiB)

julia> @btime collect(nearby_ids($pos, model, 10)) setup=(model=aperiodic_model(GridSpaceSingle))
  1.657 μs (7 allocations: 2.78 KiB)

julia> # positions
       @btime collect(nearby_positions($pos, model, 1)) setup=(model=periodic_model(GridSpace))
  37.874 ns (2 allocations: 224 bytes)

julia> @btime collect(nearby_positions($pos, model, 1)) setup=(model=aperiodic_model(GridSpace))
  38.491 ns (2 allocations: 224 bytes)

julia> @btime collect(nearby_positions($pos, model, 10)) setup=(model=periodic_model(GridSpace))
  449.297 ns (2 allocations: 7.11 KiB)

julia> @btime collect(nearby_positions($pos, model, 10)) setup=(model=aperiodic_model(GridSpace))
  414.357 ns (2 allocations: 7.10 KiB)
```

which means that the only case where this optimizations has no effect is actually GridSpace :D , but I think that this is because I did it rather blindly and I started from the other cases, and then I applied it a bit in a hurry and was fooled by wrong benchmarks

BUT it is almost 10x with a radius of 10 for `nearby_positions` !! And 2x in general for `nearby_ids` in a periodic `GridSpaceSingle`

**Comment by Tortar** [comment]
> ```julia
> (i -> r < pos[i] <= space_size[i] - r, 1:D)
> ```

If I recall correctly the positions starts from 1 and the extent of the space is inclusive so it should be right

**Comment by Tortar** [comment]
Incredibly enough, even this implementation for `GridSpace` seems to bring some benefit, I ran Flocking multiple times but using Periodic = False and in this case it becomes faster, really almost 2x in the large model, while with periodic = True as it is the correct Flocking everything is still the same speed. Don't know what to think actually, these results are strange, but didn't find any other way to use the idea to optimize the code further. So I think everything should be mergeable as it is now

File: src/spaces/grid_general.jl [hunk 1]
```diff
@@ -134,18 +134,24 @@ function nearby_positions(
     ) where {D}
     stored_ids = space.stored_ids
     nindices = get_indices_f(space, r)
-    positions_iterator = (n .+ pos for n in nindices)
-    return Base.Iterators.filter(
-        pos -> checkbounds(Bool, stored_ids, pos...), positions_iterator
-    )
+    space_size = size(stored_ids)
+    if all(i -> r < pos[i] <= space_size[i] - r, 1:D)
```
Comment (Datseris) [inline src/spaces/grid_general.jl:138]:
```suggestion
    # check if we are far from the wall to skip bounds checks
    if all(i -> r < pos[i] <= space_size[i] - r, 1:D)
```

File: src/spaces/grid_multi.jl [hunk 2]
```diff
@@ -120,40 +122,41 @@ struct GridSpaceIdIterator{P,D}
     origin::NTuple{D,Int}             # origin position nearby is measured from
     L::Int                            # length of `indices`
     space_size::NTuple{D,Int}         # size of `stored_ids`
+    nocheck::Bool
```
Comment (Datseris) [inline src/spaces/grid_multi.jl:125]:
```suggestion
    nocheck::Bool                          # skip bound checks if we are far from the edges
```

**Review by Datseris — APPROVED** [review]
Okay, so if I understand this PR right, it checks if the chebyshev distance of the cell is further away from the walls than the radius. If it is, it skips completely doing the bounds check or the mod1 operation. That makes sense, pretty stupid of us to not do it from the very first implementation.

**Comment by Datseris** [comment]
go ahead and merge after adding some comments and a changelog entry if need be

File: src/spaces/grid_multi.jl [hunk 3]
```diff
@@ -138,9 +138,8 @@ combine_positions(pos, origin, ::GridSpaceIdIterator{false}, ::Bool) = pos .+ or
 function combine_positions(pos, origin, iter::GridSpaceIdIterator{true}, nocheck)
     # the mod function is not needed for many positions and it's expensive compared
     # with checking for bounds so it is better to apply it only as a fallback.
-    off_pos = pos .+ origin
-    (nocheck || checkbounds(Bool, iter.stored_ids, off_pos...)) && return off_pos
-    return mod1.(off_pos, iter.space_size)
+    npos = pos .+ origin
+    (nocheck || checkbounds(Bool, iter.stored_ids, npos...)) ? npos : mod1.(npos, iter.space_size)
```
Comment (Datseris) [inline src/spaces/grid_multi.jl:142]: this isn't really readable. Use an `if else` statement instead.

**Comment by Tortar** [comment]
> Okay, so if I understand this PR right, it checks if the chebyshev distance of the cell is further away from the walls than the radius. If it is, it skips completely doing the bounds check or the mod1 operation. That makes sense, pretty stupid of us to not do it from the very first implementation.

that's exactly how it works :-)

**Pull request merged**